File: modules/papillon-native/src/PapillonNativeModule.ts
import { NativeModule, requireNativeModule } from "expo";

export interface SpotlightDebugSnapshot {
  activeAccountIds: string[];
  indexedCourseCount: number;
  indexedHomeworkCount: number;
  indexedGradeCount: number;
  rawCourseCount: number;
  rawHomeworkCount: number;
  rawGradeCount: number;
  lastIndexedAt: number | null;
}

declare class PapillonNativeModule extends NativeModule {
  reindexSpotlight(accountIds: string[]): Promise<void>;
  clearSpotlightIndex(): Promise<void>;
  getSpotlightDebugSnapshot(): Promise<SpotlightDebugSnapshot>;
}

export default requireNativeModule<PapillonNativeModule>("PapillonNative");

File: modules/papillon-native/ios/PapillonNativeModule.swift
import ExpoModulesCore

public final class PapillonNativeModule: Module {
    public func definition() -> ModuleDefinition {
        Name("PapillonNative")

        AsyncFunction("reindexSpotlight") { (accountIds: [String]) in
            try await PapillonKit.reindexSpotlight(accountIds: accountIds)
        }

        AsyncFunction("clearSpotlightIndex") {
            try await PapillonKit.clearSpotlightIndex()
        }

        AsyncFunction("getSpotlightDebugSnapshot") { () -> [String: Any] in
            let snapshot = await PapillonKit.debugSnapshot()
            return [
                "activeAccountIds": snapshot.activeAccountIds,
                "indexedCourseCount": snapshot.indexedCourseCount,
                "indexedHomeworkCount": snapshot.indexedHomeworkCount,
                "indexedGradeCount": snapshot.indexedGradeCount,
                "rawCourseCount": snapshot.rawCourseCount,
                "rawHomeworkCount": snapshot.rawHomeworkCount,
                "rawGradeCount": snapshot.rawGradeCount,
                "lastIndexedAt": snapshot.lastIndexedAt.map { $0.timeIntervalSince1970 * 1000 } ?? NSNull(),
            ]
        }
    }
}
